### PR TITLE
[Setup]Add table prefix to read table on schema builder

### LINKED ISSUE #33320: Cannot run "php bin/magento setup:upgrade" command when 2 stores with different table prefixes are installed in one database.
### Preconditions (*)

Magento have a default option to use a database with prefix or without prefix. Same is true that you can install multiple Magento instances into one database.
We installed one default Magento with prefix and one default Magento without prefix into one MySQL database. So we had 2 default Magento installations within one MySQL database, one with a prefix and another one without any prefix:

Store_1
Magento 2.4.2p1
PHP 7.4
Store_1/app/etc/env.php
 'db' => ['table_prefix' => 'sp2_','connection' => ['default' => ['host' => '192.168.0.31:3306','dbname' => 'magento237','username' => 'root','password' => 'root','model' => 'mysql4','engine' => 'innodb','initStatements' => 'SET NAMES utf8;','active' => '1','driver_options' => [1014 => false]]

Store_2
Magento 2.3.7
PHP 7.4
Store_2/app/etc/env.php
 'db' => ['table_prefix' => '','connection' => ['default' => ['host' => '192.168.0.31:3306','dbname' => 'magento237','username' => 'root','password' => 'root','model' => 'mysql4','engine' => 'innodb','initStatements' => 'SET NAMES utf8;','active' => '1','driver_options' => [1014 => false]]

### Steps to reproduce (*)

1. Connect to Store_1(magento with prefix) through command-line.
2. Execute the "php bin/magento setup:upgrade" command.

### Expected result (*)

It must complete the operation of the command without any error.

### Actual result (*)

Table error message:
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'magento237.sp2_catalogsearch_fulltext_scope1' doesn't exist, query was: SHOW CREATE TABLE `sp2_catalogsearch_fulltext_scope1`

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Den4ik**
### Description (*)

Prepare list of tables based on table prefix from config. It prevent issue with multiple magento installations into one database.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33320

### Manual testing scenarios (*)

See magento/magento2#33320

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Integration Tests

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests look not related to changes from this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9129](https://jira.corp.magento.com/browse/ENGCOM-9129) has been created to process this Pull Request

File: lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php [hunk 1]
```diff
@@ -225,7 +225,8 @@ public function readTables($resource)
                 ['TABLE_NAME']
             )
             ->where('TABLE_SCHEMA = ?', $dbName)
-            ->where('TABLE_TYPE = ?', self::MYSQL_TABLE_TYPE);
+            ->where('TABLE_TYPE = ?', self::MYSQL_TABLE_TYPE)
+            ->where('TABLE_NAME LIKE ?', $this->resourceConnection->getTablePrefix() . '%');
```
Comment (fredden) [inline lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php]: Should wildcard characters (like `_`) be escaped from `$this->resourceConnection->getTablePrefix()`?

Comment (Den4ik) [reply lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php]: @fredden good point. And yes, it should be escaped

**Comment by Den4ik** [comment]
@magento run all tests

**Review by fredden — APPROVED** [review]

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please fix test failures

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php [hunk 2]
```diff
@@ -226,6 +238,13 @@ public function readTables($resource)
             )
             ->where('TABLE_SCHEMA = ?', $dbName)
             ->where('TABLE_TYPE = ?', self::MYSQL_TABLE_TYPE);
+
+        if ($tablePrefix = $this->resourceConnection->getTablePrefix()) {
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php]: Could you remove variable creation in the assertion? Like this:
```suggestion
        $tablePrefix = $this->resourceConnection->getTablePrefix();
        if ($tablePrefix) {
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please fix integration test failures and do some small improvement

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9129](https://jira.corp.magento.com/browse/ENGCOM-9129) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run all tests

**Review by eduard13 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-9129](https://jira.corp.magento.com/browse/ENGCOM-9129) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
Hi @Den4ik ,

Thanks for your contribution and collaboration.

There are conflicts in the file, please resolve the conflicts. Kindly test this in the latest `2.4-develop` branch if you are facing any issues please inform us. Moving this to the Changes Requested.

Thanks

**Comment by engcom-Lima** [comment]
Hi @Den4ik ,

Kindly do the needful.

Thanks

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@engcom-Lima Conflict is resolved. You may continue with processing of PR

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Hello @Den4ik,

Thank you for your contributions!

Please resolve the conflict so we can proceed with this PR?

Thanks again!

**Comment by engcom-Dash** [comment]
Closing this PR since it has not been updated in a while. Please feel free to reopen if needed.

**Pull request closed without merging**